File: src/0438-Find-All-Anagrams-in-a-String/0438.cpp
class Solution {
 public:
  vector<int> findAnagrams(string s, string p) {
    int n = s.size(), m = p.size();
    if (n < m) return {};

    int cnt[26] = {};
    for (int i = 0; i < m; i++) cnt[p[i] - 'a']++;

    int t = m;
    vector<int> res;
    for (int i = 0; i < n; i++) {
      if (cnt[s[i] - 'a'] >= 1) t--;
      cnt[s[i] - 'a']--;
      if (i >= m) {
        if (cnt[s[i - m] - 'a'] >= 0) t++;
        cnt[s[i - m] - 'a']++;
      }
      if (t == 0) res.emplace_back(i - m + 1);
    }
    return res;
  }
};

File: src/0438-Find-All-Anagrams-in-a-String/0438.rs
impl Solution {
    pub fn find_anagrams(s: String, p: String) -> Vec<i32> {
        let (n, m) = (s.len(), p.len());
        let mut res = vec![];
        if n < m {
            return res;
        }

        let mut cnt = vec![0; 26];
        let s = s.as_bytes();
        let p = p.as_bytes();
        for i in 0..m {
            let x = (p[i] - b'a') as usize;
            cnt[x] += 1;
        }

        let mut t = m;
        for i in 0..n {
            let x = (s[i] - b'a') as usize;
            if cnt[x] >= 1 {
                t -= 1;
            }
            cnt[x] -= 1;
            if i >= m {
                let x = (s[i - m] - b'a') as usize;
                if cnt[x] >= 0 {
                    t += 1;
                }
                cnt[x] += 1;
            }
            if t == 0 {
                res.push((i - m + 1) as i32);
            }
        }
        res
    }
}
